File: Assets/Scripts/Shared/Grid2D.cs
using System.Collections.Generic;

namespace Kukumberman.Shared
{
    public sealed class Grid2D
    {
        private readonly int _width;
        private readonly int _height;

        public int Width => _width;
        public int Height => _height;

        public Grid2D(int width, int height)
        {
            _width = width;
            _height = height;
        }

        public List<int> GetNeighbours(int x, int y)
        {
            var list = new List<int>();
            GetNeighboursNonAlloc(x, y, list);
            return list;
        }

        public void GetNeighboursNonAlloc(int x, int y, List<int> indexes)
        {
            for (int i = -1; i <= 1; i++)
            {
                for (int j = -1; j <= 1; j++)
                {
                    if (i == 0 && j == 0)
                    {
                        continue;
                    }
                    var xx = x + i;
                    var yy = y + j;
                    if (IsInside(xx, yy))
                    {
                        var index = ConvertTo1D(xx, yy);
                        indexes.Add(index);
                    }
                }
            }
        }

        public bool IsInside(int x, int y)
        {
            var xx = x >= 0 && x < _width;
            var yy = y >= 0 && y < _height;
            return xx && yy;
        }

        public int ConvertTo1D(int x, int y)
        {
            return y * _width + x;
        }

        public bool ConvertTo2D(int index, out int x, out int y)
        {
            x = index % _width;
            y = index / _width;
            return IsInside(x, y);
        }
    }
}

File: react/src/core/Grid2D.ts
export default class Grid2D {
  public readonly Width: number;
  public readonly Height: number;

  public constructor(width: number, height: number) {
    this.Width = width;
    this.Height = height;
  }

  public GetNeighbours(x: number, y: number): number[] {
    const list = [];
    this.GetNeighboursNonAlloc(x, y, list);
    return list;
  }

  public GetNeighboursNonAlloc(x: number, y: number, indexes: number[]) {
    for (let i = -1; i <= 1; i++) {
      for (let j = -1; j <= 1; j++) {
        if (i === 0 && j === 0) {
          continue;
        }
        var xx = x + i;
        var yy = y + j;
        if (this.IsInside(xx, yy)) {
          const index = this.ConvertTo1D(xx, yy);
          indexes.push(index);
        }
      }
    }
  }

  public IsInside(x: number, y: number) {
    const xx = x >= 0 && x < this.Width;
    const yy = y >= 0 && y < this.Height;
    return xx && yy;
  }

  public ConvertTo1D(x: number, y: number) {
    return y * this.Width + x;
  }

  public ConvertTo2D(index: number) {
    const x = index % this.Width;
    const y = Math.floor(index / this.Width);
    return {
      x,
      y,
      inside: this.IsInside(x, y),
    };
  }
}
